src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const propertiesOfActios = [];
+  const stateCopy = JSON.parse(JSON.stringify(state));
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        addToSolutionArray(propertiesOfActios, stateCopy);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        addToSolutionArray(propertiesOfActios, stateCopy);
+        break;
+
+      case 'clear':
+        if (Object.keys(stateCopy).length === 0) {
+          addToSolutionArray(propertiesOfActios, stateCopy);
+          break;
+        } else {
+          clearProperties(stateCopy);
+          addToSolutionArray(propertiesOfActios, stateCopy);
+        }
+        break;
+    }
+  }
+
+  return propertiesOfActios;
+}
+
+function addProperties(stateCopy, extraData) {
+  Object.assign(stateCopy, extraData);
+}
+
+function removeProperties(stateCopy, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete stateCopy[key];
+  }
+}
+
+function clearProperties(stateCopy) {
+  for (const key in stateCopy) {
+    delete stateCopy[key];
+  }
```

Comment:
The function name `clearProperties` is clear, but consider using a more descriptive name for the parameter `stateCopy`, such as `targetState` or `currentState`, to indicate that this is the state being cleared.